### DISCUSSION TITLE
partialResult of customBuiltin

### DISCUSSION THREAD

**Discussion by Evange95** [open-policy-agent/community#117] [OPA and Rego]
Hello everyone, I would like to know if it is possible to partially evaluate a policy rego that uses a custom builtin.
In detail the builtin performs a call to a mongodb collection taking all the parameters needed from the context.
I'm currently having some problems with it and I want to know if it is not possible in general.

Thank you!

**Comment by srenatus** [comment]
Built-ins in general are partially evaluated as follows:
- if all their arguments are _known_, and the builtin is not declared _impure_, it will be evaluated and the result used to further evaluate the quert
- if any of their arguments are _unknown_, or the builtin is declared _impure_, a call to the builtin with the arguments (evaluated as far as possible) is _saved_ -- i.e., added to the query or support module

See the difference here, `time.now_ns()` is unsafe, and `startswith()` is not:

```
$ opa eval -fsource -p 'time.now_ns() < 10' # impure function
# Query 1
time.now_ns() < 10

$ opa eval -fsource -p 'startswith("foox", "foo"); input > 10' # inputs known, the func call is evaluated
# Query 1
input > 10

$ opa eval -fsource -p 'startswith(input, "foo")'
# Query 1
startswith(input, "foo")
```

The list of impure built-in functions is here: https://github.com/open-policy-agent/opa/blob/v0.37.1/ast/builtins.go#L274-L283
